R/messages.R

```diff
@@ -110,3 +110,85 @@ setvar <- function(core, dates, var, values, unit)
 
     invisible(NULL)
 }
+
+
+#' Fetch all the available Hector variables
+#'
+#' This function is similar to fetchvars, except it fetches all available
+#' Hector variables, including exogenous inputs (e.g., emissions & pre-industrial
+#' concentrations), parameters (e.g., alpha, beta, ECS), and output
+#' variables (e.g., concentrations & forcings).
+#'
+#' @param core Hector core object
+#' @param dates Vector of dates, optional
+#' @param scenario Scenario name, optional str
+#' @param outpath Absolute path of the output csv. Output will only be written to
+#' csv if this is given
+#' @return vars_all    Dataframe containing all Hector variables
+#' @family main user interface functions
+#' @export
+fetchvars_all <- function(core, dates=NULL, scenario=NULL, outpath=NULL) {
+
+    if (is.null(scenario)) {
+        scenario <- core$name
+    }
+
+    strt <- startdate(core)
+    end <- getdate(core)
+
+    if (is.null(dates)) {
+        dates <- strt:end
+    } else {
+        valid <- dates >= strt & dates <= end
+        dates <- dates[valid]
+    }
+    
+    # Get output for variables that do use the date param
+    rslt_date  <- do.call(rbind,
+                      lapply(vars_date, function(v) {
+                            sendmessage(core, GETDATA(), v, dates, NA, '')
+                      }))
+    
+    # Get variables that DO NOT use date arg
+    rslt_nodate <- do.call(rbind,
+                       lapply(vars_nodate, function(v) {
+                            sendmessage(core, GETDATA(), v, NA, NA, '')
+                       }))
+    
+    rslt <- rbind(rslt_date, rslt_nodate)
+    rslt$scenario <- scenario
+    
+    if (!is.null(outpath)) {
+
+        # Pivot the dataframe "wide" so years run along the x axis
+        # Catch warning from reshape()
+        suppressWarnings(rslt_pretty <- reshape(rslt, direction="wide",
+                                        idvar=c("scenario", "variable", "units"),
+                                        timevar="year"))
+
+        # Clean up the column names (value.YYYY --> YYYY)
+        col_names <- colnames(rslt_pretty)[-1:-3]
+        col_names <- sapply(col_names, function(x) sub("value.", "", x), USE.NAMES=F)
+        col_names <- col_names[-length(col_names)]
+
+        col_names <- c("scenario", "variable", "units", "initial_value", col_names)
+
+        num_cols <- length(col_names)
+
+        # Move last column (initial_value) into 4th col position & scenario into 1st col
+        rslt_pretty <- rslt_pretty[, c(3, 1, 2, num_cols, 5:num_cols-1)]
+
+        # Set the column names for the re-ordered dataframe
+        colnames(rslt_pretty) <- col_names
+
+        # Construct output filename and absolute path
+        f_name <- "fetchvars_all.csv"
+        abs_path <- file.path(outpath, f_name)
```

Comment:
Personal preference, but I would just make `outpath` the output file name (you could change to `outfile` to make that clearer). I could imagine users wanting to write out multiple files from different Hector simulations to the same directory.

Reply:
Good idea...I'm going to add a snippet that checks for the `.csv` extension at the end of `outfile`. 
Would it be better to stop execution if its not there or raise a warning, append `.csv` to `outfile`, and continue?

Reply:
> checks for the .csv extension at the end of outfile

Personally, I wouldn't. Sometimes, you might want a different suffix (e.g. `.dat`, `.txt) or no suffix at all, and I tend to prefer such things not be forced on people.

My preference would be to pass arguments to `write.table`, with common arguments for `csv` as the default. I.e. Something like:

```r
write_hector_table <- function(x, file, sep = ",", row.names = FALSE, col.names = TRUE, ...) {
  pretty_table <- make_pretty(x) # all the stuff you did
  # Provide different default values for `sep`, `row.names`, and `col.names` so the default format is CSV
  # But, because they are arguments, they can still be modified if the user wants
  # Also, the <...> means "pass any other named argument down to write.table
  write.table(pretty_table, file, sep = sep, row.names = row.names, col.names = col.names, ...)
}
```

With that function, the user has a lot of flexibility about the output format. E.g.:

```r
# Tab-separated file
write_hector_table(results, "mydata.tsv", sep = "\t")
# CSV with row names
write_hector_table(results, "mydata.csv", row.names = TRUE)
# CSV without column names and a different NA value
# NOTE -- The `...` allows write_hector_table to inherit the `na` arg from write.table
write_hector_table(results, "mydata.csv", col.names = FALSE, na = -9999)
```